### ISSUE TITLE
How to configure .net application running in kubernetes to see the app in the newrelic APM tab?

### ISSUE THREAD

**Issue by Amit-limbasiya** [newrelic/newrelic-dotnet-agent#2567]
I have the application running inside the kubernetes pod.

I have installed the newrelic dotnet agent in the kubernetes daemonset via helm command shown in the offical documentation.

Now I want that this .net agent will identify the dotnet application running inside the pod and forward its log as well as APM data like avg response time, % of error etc.

Currently when the application starts i.e. pod is up and running, it is sending the logs in the kubernetes section in newrelic account. where I can see the logs and filter them by pod name or container name. But I cant see the APM data or application in the APM tab. How/what to configure to see the APM data of the application.

New relic dotnet agent installed in the kubernetes daemonset.

I have added two nuget packages in the application:
```
<PackageReference Include="Serilog.AspNetCore" Version="8.0.1" />
<PackageReference Include="Serilog.Sinks.NewRelic" Version="2.0.0" />
```

In the launchSettings.json I have added the following environment variables:
```
"ASPNETCORE_ENVIRONMENT": "Development",
"CORECLR_ENABLE_PROFILING": "1"
```

I have configured the serilog as following in the appsettings.json:
```
"Serilog": {
"Properties": {
"app": "test_api",
"Application": "test_api"
},
"WriteTo": {
"NewRelicLogsSink": {
"Name": "NewRelic",
"Args": {
"applicationName": "test_api"
}
}
}
},
```

Create and Use the log in program.cs:
Creation:
```
var logger = new LoggerConfiguration()
.ReadFrom.Configuration(builder.Configuration, "Serilog")
.Enrich.FromLogContext()
.CreateLogger();

builder.Logging.ClearProviders();
builder.Logging.AddSerilog(logger);

```
Use:
`Log.Information("This is get endpoint from test api");`

**pod.yaml** file
```
apiVersion: v1
kind: Pod
metadata:
  name: test-api-serilog-pod
  labels:
    app: test-api-serilog
spec:
  containers:
    - name: test-api-serilog-container
      image: test-api-serilog:latest
      imagePullPolicy: Never
      ports:
      - containerPort: 8080
      env:
      - name: ASPNETCORE_ENVIRONMENT
        value: "Development"
      - name: CORECLR_ENABLE_PROFILING
        value: "1"
      - name: NEW_RELIC_APP_NAME
        value: "A new test api serilog 2"        
      - name: NEW_RELIC_LOG
        value: "stdout"
      - name: NEW_RELIC_APPLICATION_LOGGING_ENABLED
        value: "true"
      - name: NEW_RELIC_APPLICATION_LOGGING_FORWARDING_ENABLED
        value: "true"
      - name: NEW_RELIC_LICENSE_KEY
        valueFrom:
          secretKeyRef: 
            name: newrelic-bundle-newrelic-logging-config
            key: license
```

I am able to get logs on the newrelic cloud, which I can filter by `pod_name` or `container_name`. 

**Expecting:**
Now I want the APM Data in the newrelic cloud like avg.response time or % of error, type of error etc. 
[![APM ss](https://i.sstatic.net/Wi2o9VBw.png)](https://i.sstatic.net/Wi2o9VBw.png)
Can't see the A new test api serilog 2 (Value of NEW_RELIC_APP_NAME) in the APM tag.
Additionally Inside the newrelic cloud **Kubernetes->pods->test-api-serilog-pod** I can see the logs of the pod but can't see the APM data.

**Doubt:**
**Is it necessary to have the dotnet agent installed inside the application running container, Or can we configure the dotnet agent running in daemonset to send the APM data to newrelic as well?**

**Comment by tippmar-nr** [comment]
Hi @Amit-limbasiya - Thanks for opening this issue. 

As it happens, New Relic is launching the public preview of our new Kubernetes Operator that will greatly simplify installation of any of our language agents in your Kubernetes cluster. 

The preview is scheduled to go live on June 27, so please watch the [New Relic blog](https://newrelic.com/blog) for an announcement. If you want to see what's coming, you can take a look at our [Kubernetes Operator repo](https://github.com/newrelic/k8s-agents-operator) (but keep in mind that the agent init container images are not publicly available yet).

**Comment by tippmar-nr** [comment]
@Amit-limbasiya Just a quick followup - our Kubernetes Operator is now in public preview; please reference the documentation at https://docs.newrelic.com/docs/kubernetes-pixie/kubernetes-integration/installation/k8s-agent-operator/ if you'd like to try it out.

**Comment by Amit-limbasiya** [comment]
Sure, I will try it out. Thanks @tippmar-nr for letting me know.

**Comment by Amit-limbasiya** [comment]
Hi @tippmar-nr, I have tried it out. With the use of the kubernetes agent operator along with the kubernetes integration I can get the log-forwarding and APM data enabled and can get the data in new relic ui.

I have one question: Is the agent production-ready? If I have 100s microservices and I remove the agent from container(usually done by the installation commands in dockerfile), and just keep one kubernetes agent operator and kubernetes integration, Can I achieve same results? 

Secondly, What is the difference to add the newrelic dotnet agent in container via installation commands in dockerfile and this kubernetes agent operator with newrelic kubernetes integration, in terms of only dotnet applications. 
I see one difference when the different types of applications are deployed agents with integration gives the flexibility. But in the scenario where I have 100s of microservices of only only dotnet type, Will this be beneficial to remove agents from each microservices and install the kubernetes integration with kubernetes agent operator?

**Comment by tippmar-nr** [comment]
@Amit-limbasiya I'm happy to hear that you've been able to get the agent operator working in your environment. Thanks for trying it out! 

Even though the k8s operator is in public preview, the agent installed by the operator is our latest release, so it's definitely production ready once the operator has installed the agent. 

The agent needs to be installed in each container where you want APM data to be collected and reported. 

If you have many services that all currently install the agent in `dockerfile`, using the k8s agent operator in your pod spec would be an ideal way to decouple the APM agent dependency from your individual services - the operator will ensure that every container in the pod is running the same version of the agent and requires you to only update a single file when you want to upgrade to a newer version of the agent. 

Note that if a container installs the .NET agent via `dockerfile` and is later referenced in a pod spec that uses the agent operator, the existing installation of the agent (that is, the one installed via `dockerfile`) will take precedence and the agent operator will not install the agent. 

If you have additional questions, I would encourage you to reach out to [New Relic Support](https://support.newrelic.com/s/) for further assistance.

**Comment by Amit-limbasiya** [comment]
Thank you @tippmar-nr, No more doubts now,
Just to confirm, 
The agent installed via `dockerfile` is the old way to install agent and have dependency in dockerfile with application.
k8s operator agent will do the same with out newrelic agent's dependency in dockerfile, **it will also install** the agent in each application running container but in the decoupled way. Now we have to manage only operator to update the version of the agent running in all many services. And it will install the agent by running an init container (as I can see it by describing the pod), right?

**Comment by tippmar-nr** [comment]
@Amit-limbasiya Yes, you are understanding how the k8s agent operator works - you no longer need to install the agent in `dockerfile`; just add the operator init container to your deployment spec and the agent will be installed automatically.

**Comment by Amit-limbasiya** [comment]
Hii @tippmar-nr, sorry for asking questions here, actually I couldnt find any discussion area where I can discuss the question. So dropping it here.

I am using kubernetes agent operator. It is running pod for bundles and etc. in newrelic namespace. I have configured the production namespace in the agent operator to install newrelic agent in the application running inside the production namespace. So I want to ask that, now if there is currently one node and 3 replicas of application is running and I can see the APM data of those replicas in newrelic. Now my auto scallar detect more traffic and there is no space in current node so it automatically scale one additional node(not pod) and deploy application in production namespace in that node, then will it configure the apm agent to install newrelic agent for the new node?

Basically, will the agent operator works fine in the case of node auto scalling? or any other additional things/installation of agent operator needs to be done when the new node appears?

**Comment by tippmar-nr** [comment]
Hey @Amit-limbasiya - The k8s agent operator should work exactly as expected in an auto-scaling scenario, whether that's adding a new pod within the node or when bringing a new node online.

For future reference, you can ask questions like this on the [New Relic Explorer's Hub](https://forum.newrelic.com/s/)
